src/query/functions/binary/node.go

```diff
@@ -0,0 +1,482 @@
+func (o binaryOp) OpType() string {
+	return o.OperatorType
+}
+
+// String representation
+func (o binaryOp) String() string {
+	return fmt.Sprintf("type: %s, lnode: %s, rnode: %s", o.OpType(), o.info.LNode, o.info.RNode)
+}
+
+// Node creates an execution node
+func (o binaryOp) Node(controller *transform.Controller) transform.OpNode {
+	return &binaryNode{
+		controller: controller,
+		cache:      transform.NewBlockCache(),
+		op:         o,
+	}
+}
+
+type binaryNode struct {
+	op         binaryOp
+	controller *transform.Controller
+	cache      *transform.BlockCache
+	mu         sync.Mutex
+}
+
+// NodeInformation describes the types of nodes used
+// for binary operations
+type NodeInformation struct {
+	LNode, RNode         parser.NodeID
+	LIsScalar, RIsScalar bool
+	ReturnBool           bool
+	VectorMatching       *logical.VectorMatching
+}
+
+// NewBinaryOp creates a new binary operation
+func NewBinaryOp(
+	opType string,
+	info NodeInformation,
+) (parser.Params, error) {
+	fn, ok := mathFuncs[opType]
+	if !ok {
+		return binaryOp{}, fmt.Errorf("operator not supported: %s", opType)
+	}
+
+	return binaryOp{
+		OperatorType: opType,
+		// LNode:        lNode,
+		// RNode:        rNode,
+		fn:   fn,
+		info: info,
+	}, nil
+}
+
+// Process processes a block
+func (n *binaryNode) Process(ID parser.NodeID, b block.Block) error {
+	lhs, rhs, err := n.computeOrCache(ID, b)
+	if err != nil {
+		// Clean up any blocks from cache
+		n.cleanup()
+		return err
+	}
+
+	// Both blocks are not ready
+	if lhs == nil || rhs == nil {
+		return nil
+	}
+
+	n.cleanup()
+
+	nextBlock, err := n.process(lhs, rhs)
+	if err != nil {
+		return err
+	}
+
+	defer nextBlock.Close()
+	return n.controller.Process(nextBlock)
+}
+
+// processes two logical blocks, performing a logical operation on them
+func (n *binaryNode) process(lhs, rhs block.Block) (block.Block, error) {
+	info := n.op.info
+
+	lIter, err := lhs.StepIter()
+	if err != nil {
+		return nil, err
+	}
+
+	fn := n.op.fn
+
+	if info.LIsScalar {
+		scalarL, ok := lhs.(*block.ScalarBlock)
+		if !ok {
+			return nil, errLeftScalar
+		}
+
+		lVal := scalarL.Value()
+		// if both lhs and rhs are scalars, can create a new block
+		// by extracting values from lhs and rhs instead of doing
+		// by-value comparisons
+		if info.RIsScalar {
+			scalarR, ok := rhs.(*block.ScalarBlock)
+			if !ok {
+				return nil, errRightScalar
+			}
+
+			// NB(arnikola): this is a sanity check, as scalar comparisons
+			// should have previously errored out during the parsing step
+			if !n.op.info.ReturnBool {
+				return nil, errNoModifierForComparison
+			}
+
+			return block.NewScalarBlock(
+				fn(lVal, scalarR.Value()),
+				lIter.Meta().Bounds,
+			), nil
+		}
+
+		// rhs is a series; use rhs metadata and series meta
+		return n.processSingleBlock(
+			rhs,
+			func(x float64) float64 {
+				return n.op.fn(lVal, x)
+			},
+		)
+	}
+
+	if info.RIsScalar {
+		scalarR, ok := rhs.(*block.ScalarBlock)
```

Comment:
some duplication here:

maybe create a function which takes two block.Block, assumes first can be casted into scalar and other can't be. Then you can use the same function for this case and one of the previous cases